events/api/src/main/java/org/projectnessie/events/api/MergeEvent.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (C) 2023 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.events.api;
+
+import com.fasterxml.jackson.annotation.JsonTypeName;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import com.google.errorprone.annotations.CanIgnoreReturnValue;
+import org.eclipse.microprofile.openapi.annotations.enums.SchemaType;
+import org.eclipse.microprofile.openapi.annotations.media.Schema;
+import org.immutables.value.Value;
+
+/**
+ * Event that is emitted when a merge is performed. This event is emitted after the merge has been
+ * persisted.
+ */
+@Schema(
+    type = SchemaType.OBJECT,
+    description =
+        "An event that is emitted when a merge is performed. "
+            + "This event is emitted after the merge has been persisted.")
+@Value.Immutable
+@JsonTypeName("MERGE")
+@JsonSerialize
+@JsonDeserialize
+public interface MergeEvent extends CommittingEvent {
```

Comment:
Note that, as agreed, `MergeEvent` and `TransplantEvent` do not include the individual commits that were added to the target branch. This is to prevent huge event payloads, but the resulting objects are rather tiny now.

I wonder if we could improve this.

Assume we have a merge operation that appended 10 commits to the target branch, containing 50 operations in total.

We currently would emit:

1. One `MergeEvent`;
2. 50 `ContentStoredEvent` or `ContentRemovedEvent` events;

That is, 51 events total.

We could instead emit:

1. One `MergeEvent`;
2. 10 `CommitEvent`s;
2. 50 `ContentStored` or `ContentRemoved` events.

That is, 61 events total.

The advantage of this is that individual commit events would be sent along will their `CommitMeta` objects.

The drawback is the "write amplification" phenomenon that gets even more exacerbated.

Reply:
API v2 does not permit unquashed merges (and API v1 is supposed to be phasing out of use), so the end result will be very similar: 1 merge event, 1 commit, 50 contents. So, I'd say let's do that.

Multiple commits are possible with transplants, but I would not call it write "amplification" in that case as it would be mostly equivalent to manually re-committing the same changes on the target branch.

Reply:
FYI https://github.com/projectnessie/nessie/pull/6672